### DISCUSSION TITLE
I'ld like to read from a M-Bus Ethernet Gateway

### DISCUSSION THREAD

**Discussion by mircsicz** [wmbusmeters/wmbusmeters#993] [Ideas]
Hi all,

just stumbled upon this awesome project and am wondering if the above is possible. From a short flyover I could only see serial interfaces, so I thought let's just ask and get in touch...

I've been doing AMR since 2005 but gotten a bit rusty in the last years. My last contact was libmbus which I did setup to import stuff into MySQL.

Greetz
Mirco

**Comment by mircsicz** [comment]
If someone is willing to get this going, I'ld love to but am no dev I'll happily supply a m-bus ethernet gateway ;-) I've got three on the shelf...

**Comment by mircsicz** [comment]
I might have found a solution but have yet to approve it: [socat](http://www.dest-unreach.org/socat)

Prob is I'm not getting packages / data through...

`socat PTY,link=/dev/csTTY1,echo=0,wait-slave TCP:10.10.1YY.XX:10001`

just doesn't give feedback nor anything from the other side of the tunnel

`socat PTY,link=/dev/csTTY1,echo=0,wait-slave TCP:10.10.1xx.xx:10001`
won't even start:
2023/07/14 01:18:54 socat[15178] E tcgetattr(5, 0x9fb74): Inappropriate ioctl for device

Both are suggestions: 
- [1](https://www.wut.de/e-58www-29-apus-000.php)
- [2](https://gist.github.com/DraTeots/e0c669608466470baa6c)

Hoping for a hint

BTW: M-Bus and Master are fine, see screenshot
![2023-07-14 02_01_40-NG-MScan 2 (2 0 039)](https://github.com/wmbusmeters/wmbusmeters/assets/7398305/3d9a3942-8626-435e-8f10-94a4e72b2a68)

**Comment by weetmuts** [comment]
Do you have some documentation on the  m-bus ethernet gateway?
In particular how do you set the serial speed of the m-bus signals?

**Reply by mircsicz** [reply]
@weetmuts thanks for your reply. The device is a M-Bus 810 from PiiGAB, here's the links to their documentation. I've just talked to one of their devs on the phone and he told me that one of their dev's is quite Linux aware. I'll send an email to their support and ask for his help too! I'm quite sure it's just a question of using the correct socat params... 

[Productpage](https://www.piigab.com/product/piigab-m-bus-810-m-bus-omvandlare/)
[Datasheet](https://www.piigab.com/download/piigab-m-bus-810-data-sheet-eng/?wpdmdl=60560&refresh=64b11c91128cb1689328785)
[Docs](https://www.piigab.com/download/piigab-m-bus-810-manual-eng/?wpdmdl=60557&refresh=64b11c912a9571689328785)

There's also swedish versions of their docs on the Productpage ;-)

Regarding your question: I'm pretty sure theirs no need to set the bus-speed, when reading it on windows the software [MBConf](https://service.relay.de/media/software/MBConf_Setup.exe) ran with 2400 and there's no setting on the TCP-serial driver. And I just tested with another app too, see the attached screenshots.

![2023-07-14 13_03_26-LorusFree Version 2023                           LorusA dll version V2022-10-12 ](https://github.com/wmbusmeters/wmbusmeters/assets/7398305/72f8c169-61f0-49fa-83ed-814727cb7836)
![2023-07-14 01_47_44-RELAY MBCONF  M-Bus Parametrierung  V3 8](https://github.com/wmbusmeters/wmbusmeters/assets/7398305/de0baaf3-77d3-4c7c-8921-b257c8c0f53b)

**Reply by mircsicz** [reply]
Here's another Screenshot showing the Settings for the virtual Com-Port.
![2023-07-14 13_10_01-Serial to Ethernet Connector](https://github.com/wmbusmeters/wmbusmeters/assets/7398305/ffe5af59-4b51-404b-b613-e662ca3d58b9)

**Comment by mircsicz** [comment]
So here's the solution:

`socat -d -d pty,rawer,echo=0,b2400 tcp:10.10.160.217:10001`

And running [libmbus](http://www.rscada.se/libmbus/index.php?lang=en) proved that it works… The rest is now up to the project maintainer of wmbusmeters

`mbus-serial-request-data -d /dev/pts/2 4
`
```
init_slaves: debug: sending init frame #1                                                                                                                                                            
[2023-07-14 12:08:07Z] SEND (005): 10 40 FD 3D 16                                                                                                                                                    
init_slaves: debug: sending init frame #2                                                                                                                                                            
[2023-07-14 12:08:07Z] SEND (005): 10 40 FD 3D 16                                                                                                                                                    
[2023-07-14 12:08:07Z] SEND (005): 10 5B 04 5F 16                                                                                                                                                    
[2023-07-14 12:08:08Z] RECV (159): 68 99 99 68 08 04 72 04 00 00 00 2E 0F 00 00 14 02 00 00 04 03 11 00 00 00 04 03 00 00 00 00 04 03 00 00 00 00 04 00 00 00 00 00 04 00 00 00 00 00 04 00 00 00 00 
00 04 00 00 00 00 00 04 00 00 00 00 00 04 6D 1E 11 DD 84 04 6D 0F 11 DD 84 04 6D 1D 11 DD 84 04 20 8E 87 9C 00 01 7A 04 01 6F 00 04 6F 22 00 E2 04 0F 0D 20 10 00 01 51 03 00 01 32 10 00 01 53 03 00
 01 34 00 00 01 55 03 00 01 36 00 00 01 57 03 00 01 18 13 00 01 59 13 00 01 4A 10 00 01 4B 10 00 01 6D 04 00 01 5F 16                                                                                
mbus_frame_print: Dumping M-Bus frame [type 4, 159 bytes]: 68 99 99 68 08 04 72 04 00 00 00 2E 0F 00 00 14 02 00 00 04 03 11 00 00 00 04 03 00 00 00 00 04 03 00 00 00 00 04 00 00 00 00 00 04 00 00 
00 00 00 04 00 00 00 00 00 04 00 00 00 00 00 04 00 00 00 00 00 04 6D 1E 11 DD 84 04 6D 0F 11 DD 84 04 6D 1D 11 DD 84 04 20 8E 87 9C 00 01 7A 04 01 6F 00 04 6F 22 00 E2 04 0F 0D 20 10 00 01 51 03 00
 01 32 10 00 01 53 03 00 01 34 00 00 01 55 03 00 01 36 00 00 01 57 03 00 01 18 13 00 01 59 13 00 01 4A 10 00 01 4B 10 00 01 6D 04 00 01 5F 16 
```

So final question: How do I query a wired M-Bus and read the data from the meters found?

**Comment by mircsicz** [comment]
As a reply to myself from the [neighboring discussion](https://github.com/orgs/wmbusmeters/discussions/842) this is how I should be able to read the meter behind a wired gateway targeting its secondary address:
`
wmbusmeters --pollinterval=60s --logtelegrams MAIN=/dev/pts/2:mbus:2400 NZR em24:MAIN:mbus 090003AF434Cxxxx NOKEY
`
But all I get is:
`mircsicz@pi8:~ $ wmbusmeters --pollinterval=60s --logtelegrams MAIN=/dev/pts/2:mbus:2400 NZR em24:MAIN:mbus 090003AF434C1202 NOKEY

Started config MAIN=mbus on /dev/pts/2`
